File: PostsAndComments/app/src/main/java/com/example/postscommentsapp/di/UseCaseModule.java
package com.example.postscommentsapp.di;

import dagger.Module;
import dagger.Provides;
import dagger.hilt.InstallIn;
import dagger.hilt.components.SingletonComponent;

@Module
@InstallIn(SingletonComponent.class)
public class UseCaseModule {

    @Provides
    public GetPostsUseCase provideGetPostsUseCase(PostsRepository repo) {
        return new GetPostsUseCase(repo);
    }

    @Provides
    public SearchPostsUseCase provideSearchPostsUseCase(PostsRepository repo) {
        return new SearchPostsUseCase(repo);
    }

    @Provides
    public GetCommentsUseCase provideGetCommentsUseCase(CommentsRepository repo) {
        return new GetCommentsUseCase(repo);
    }

    @Provides
    public InsertCommentUseCase provideInsertCommentUseCase(CommentsRepository repo) {
        return new InsertCommentUseCase(repo);
    }
}

File: PostsAndComments/app/src/main/java/com/example/postscommentsapp/di/UseCaseModule.kt
package com.example.postscommentsapp.di

import com.example.postscommentsapp.domain.repository.CommentsRepository
import com.example.postscommentsapp.domain.repository.PostsRepository
import com.example.postscommentsapp.domain.usecase.*
import dagger.Module
import dagger.Provides
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent

@Module
@InstallIn(SingletonComponent::class)
object UseCaseModule {

    @Provides
    fun provideGetPostsUseCase(repo: PostsRepository) =
        GetPostsUseCase(repo)

    @Provides
    fun provideSearchPostsUseCase(repo: PostsRepository) =
        SearchPostsUseCase(repo)

    @Provides
    fun provideGetPostByIdUseCase(repo: PostsRepository) =
        GetPostByIdUseCase(repo)

    @Provides
    fun provideGetCommentsUseCase(repo: CommentsRepository) =
        GetCommentsUseCase(repo)

    @Provides
    fun provideInsertCommentUseCase(repo: CommentsRepository) =
        InsertCommentUseCase(repo)
}
